File: C++/PlusOne.cc
// Given a non-negative number represented as an array of digits, plus one to the number.

// The digits are stored such that the most significant digit is at the head of the list.

class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        int i = digits.size() - 1;
        while (i >= 0 ) {
            if (digits[i] != 9) break;
            i--;
        }
        if (i < 0) {
            vector<int> output{1};
            output.insert(output.end(), digits.size(), 0);
            return output;
        } else {
            vector<int> output(digits.begin(), digits.begin() + i + 1);
            output.insert(output.end(), digits.size() - i - 1, 0);
            output[i]++;
            return output;
        }
    }
};

File: Java/PlusOne.java
// Given a non-negative number represented as an array of digits, plus one to the number.

// The digits are stored such that the most significant digit is at the head of the list.

public class PlusOne {
    public int[] plusOne(int[] digits) {
        int carry = 1;
        for (int i = digits.length - 1; i >= 0; i--) {
            int sum = carry + digits[i];
            digits[i] = sum % 10;
            carry = sum / 10;
            if (carry == 0) {
                break;
            }
        }
        if (carry == 0) {
            return digits;
        } else {
            int[] des=new int[digits.length+1];
            des[0] = 1;
            System.arraycopy(digits, 0, des, 1, digits.length);
            return des;
        }
    }
}
